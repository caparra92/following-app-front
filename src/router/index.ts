import { useStore } from "../stores/store";
import { createRouter, createWebHistory } from '@ionic/vue-router';
import { RouteRecordRaw } from 'vue-router';
import RegisterUserPage from '../views/RegisterUserPage.vue'
import LoginPage from '../views/LoginPage.vue';
import DashboardPage from '../views/DashboardPage.vue';
import ActivitiesPage from '../views/ActivitiesPage.vue';

const loggedIn = (to: any, from: any, next: any) => {
  const store = useStore();
  if(store.loggedIn) {
    next()
    return
  }
  next('/')
}

const routes: Array<RouteRecordRaw> = [
  {
    path: '/',
    redirect: '/login'
  },
  {
    path: '/login',
    name: 'Login',
    component: LoginPage
  },
  {
    path: '/register',
    name: 'Register',
    component: RegisterUserPage
  },
  {
    path: '/dashboard',
    name: 'Dashboard',
    component: DashboardPage,
    beforeEnter: loggedIn,
    meta: {
      requiresAuth: true,
    },
    children: [ 
      { 
        path: '/dashboard/activities', 
        name: 'Activities',
        component: ActivitiesPage,
        beforeEnter: loggedIn,
        meta: {
          requiresAuth: true
        }
      }
    ]
  }
]

const router = createRouter({
  history: createWebHistory(process.env.BASE_URL),
  routes
});

router.beforeEach((to, from, next) => {
  const store = useStore();
  if (to.name !== 'Login' && !loggedIn) next({ name: 'Login' })
  else next()
});

export default router
