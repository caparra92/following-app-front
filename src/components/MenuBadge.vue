<template>
  <ion-menu content-id="main-content" size="large" class="menu-container">
    <ion-header>
      <ion-toolbar color="secondary">
        <ion-title><ion-icon :icon="statsChartOutline" class="icon"></ion-icon></ion-title>
      </ion-toolbar>
    </ion-header>
    <ion-content class="ion-padding">
      <ul class="menu-list">
        <router-link to="/dashboard"><li><ion-icon :icon="listOutline" size="large"></ion-icon>Categories</li></router-link>
        <router-link to="/register"><li><ion-icon :icon="personAddOutline" size="large"></ion-icon>Register</li></router-link>
        <li @click="logout"><ion-icon :icon="powerOutline" size="large"></ion-icon>Logout</li>
      </ul>
    </ion-content>
  </ion-menu>
  <ion-page id="main-content">
    <ion-header>
      <ion-toolbar>
        <ion-buttons slot="start">
          <ion-menu-button></ion-menu-button>
        </ion-buttons>
        <template v-if="$route.name != 'dashboard'">
          <go-back-button class="back-button"></go-back-button>
        </template>
      </ion-toolbar>
    </ion-header>
  </ion-page>
</template>

<script setup lang="ts"> 
  import router from "../router";
  import { useStore } from "@/stores/store";
  import GoBackButton from "./GoBackButton.vue";
  import {
    IonContent,
    IonHeader,
    IonMenu,
    IonMenuButton,
    IonButtons,
    IonTitle,
    IonToolbar,
    IonPage,
    IonIcon
  } from '@ionic/vue';
  import { listOutline, personAddOutline, powerOutline, statsChartOutline } from 'ionicons/icons';
  
  const store = useStore();
  
  const logout = async() => {
    const data = await store.logout();
    router.push('/login');
}
</script>
<style scoped>

.menu-container {
  position: fixed;
  overflow: scroll;
  z-index: 9999;
}
.menu-list {
  font-size: 30px;
  margin: 0;
  padding: 0;
  width: 100%;
  display: flex;
  flex-direction: column;
  text-align: center;
  border: none;
}

.menu-list li {
  display: flex;
  justify-content: space-between;
  width: 100%;
  height: 3em;
  list-style: none;
  padding: 8px;
  line-height: 1.2;
  cursor: pointer;
}

.menu-list a {
  text-decoration: none;
  font-size: .8em;
}

li ion-icon {
  padding-top: 8px;
}

.icon {
  padding-top: 10px;
}

.back-button {
  position: absolute;
  right: 10px;
  top: 5px;
}
</style>