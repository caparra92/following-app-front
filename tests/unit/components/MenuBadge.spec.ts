import { mount, VueWrapper } from "@vue/test-utils";
import { createTestingPinia } from '@pinia/testing';
import router from "@/router";
import MenuBadge from "@/components/MenuBadge.vue";

describe('MenuBadge tests', () => {

    let wrapper: VueWrapper;

    beforeEach(() => {
        const mockRoute = {
            name: "login"
          }

        const mockRouter = {
            push: jest.fn()
        }

        wrapper = mount(MenuBadge, {
            global: {
                plugins: [createTestingPinia(), router],
                mocks: {
                    $route: mockRoute,
                    $router: mockRouter
                  }
              },
        });
    });

    test('Should match with snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('Should have <ion-menu/> tag with size attribute', () => {
        const ionMenu = wrapper.find('ion-menu');
        expect(ionMenu.attributes().size).toBe("large");
    });

    test('Should have <ion-icon/> tag with class = "icon"', () => {
        const ionIcon = wrapper.findAll('ion-icon');
        expect(ionIcon[0].attributes().class).toBe("icon");
    });

    test('Should have <ion-content/> tag with class = "ion-padding"', () => {
        const ionContent = wrapper.find('ion-content');
        expect(ionContent.attributes().class).toBe("ion-padding");
    });

    test('Should have <go-back-button/> / <div> tag with class = "back-button"', () => {
        const goBackButton = wrapper.findComponent('.back-button');
        expect(goBackButton.attributes().class).toBe("back-button");
    });
});