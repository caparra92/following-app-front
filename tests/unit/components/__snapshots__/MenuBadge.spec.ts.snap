// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuBadge tests Should match with snapshot 1`] = `
<ion-menu class="" size="large">
  <ion-header class="">
    <ion-toolbar class="">
      <ion-title class="">
        <ion-icon class="icon"></ion-icon>
      </ion-title>
    </ion-toolbar>
  </ion-header>
  <ion-content class="ion-padding">
    <ul class="menu-list"><a href="/dashboard" class="">
        <li>
          <ion-icon></ion-icon>Categories
        </li>
      </a><a href="/register" class="">
        <li>
          <ion-icon></ion-icon>Register
        </li>
      </a>
      <li>
        <ion-icon></ion-icon>Logout
      </li>
    </ul>
  </ion-content>
</ion-menu>
<div id="main-content" class="ion-page">
  <ion-header class="">
    <ion-toolbar class="">
      <ion-buttons class="" slot="start">
        <ion-menu-button class=""></ion-menu-button>
      </ion-buttons>
      <div class="back-button"><button>
          <ion-icon></ion-icon>
        </button></div>
    </ion-toolbar>
  </ion-header>
</div>
`;
